### PR TITLE
PR: Add PyTorch FFmpeg to wheel and conda distributions

### PR DESCRIPTION
**Pull request by andfoy**
This PR enables the inclusion of FFmpeg LGPL binaries on wheel distributions and also updates the conda recipe dependencies to include FFmpeg as well

Fixes https://github.com/pytorch/vision/issues/2365

### CODE REVIEW

**Comment by andfoy** [comment]
@fmassa, it seems that the decoder backend is broken in Windows, is this expected?

**Comment by fmassa** [comment]
@andfoy this is not expected, and I believe we at some point tested the Windows compilation of the decoder

**Comment by andfoy** [comment]
It seems that `--disable--mmx` flag when compilling FFmpeg fixes the segfaults on the tests

**Comment by fmassa** [comment]
> It seems that --disable--mmx flag when compilling FFmpeg fixes the segfaults on the tests

So this means that we would need to ship ffmpeg without MMX optimizations? Do you have an idea if it would make the binaries much slower than the ones we currently have for some machines?

**Comment by andfoy** [comment]
> So this means that we would need to ship FFmpeg without MMX optimizations? Do you have an idea if it would make the binaries much slower than the ones we currently have for some machines?

It seems that on modern architectures MMX was superseded by SSE/2/3/4 and AVX, so in principle, disabling MMX should only affect very old processors, as FFmeg has individual flags to disable SSE and AVX: https://github.com/FFmpeg/FFmpeg/blob/05c8d0bce64888c5312822fbc9cdb63934b86519/configure#L430

File: packaging/pkg_helpers.bash [hunk 1]
```diff
@@ -303,3 +303,38 @@ setup_junit_results_folder() {
     export CONDA_PYTORCH_BUILD_RESULTS_DIRECTORY="${SOURCE_ROOT_DIR}/build_results/results.xml"
   fi
 }
+
+
+download_copy_ffmpeg() {
+  mkdir ffmpeg_tmp
+  cd ffmpeg_tmp
+  if [[ "$OSTYPE" == "msys" ]]; then
+    conda install -yq ffmpeg -c pytorch
```
Comment (peterjc123) [inline packaging/pkg_helpers.bash:312]: It is a bit weird that you pinned the version of `ffmpeg` in the following line but not in this one.

File: packaging/pkg_helpers.bash [hunk 2]
```diff
@@ -229,7 +229,7 @@ setup_pip_pytorch_version() {
 # You MUST have populated PYTORCH_VERSION_SUFFIX before hand.
 setup_conda_pytorch_constraint() {
   if [[ -z "$PYTORCH_VERSION" ]]; then
-    export CONDA_CHANNEL_FLAGS="-c pytorch-nightly"
+    export CONDA_CHANNEL_FLAGS="-c pytorch-nightly -c pytorch"
```
Comment (peterjc123) [inline packaging/pkg_helpers.bash:232]: Why do we need `pytorch` here?

File: packaging/torchvision/conda_build_config.yaml [hunk 3]
```diff
@@ -1,3 +1,5 @@
+channel_sources:
```
Comment (peterjc123) [inline packaging/torchvision/conda_build_config.yaml:1]: So we could just get rid of the arguments when we call `conda build`, right?

File: setup.py [hunk 4]
```diff
@@ -346,15 +347,16 @@ def get_extensions():
...
                     'avcodec',
                     'avformat',
                     'avutil',
                     'swresample',
                     'swscale',
                 ],
-                extra_compile_args=["-std=c++14"],
-                extra_link_args=["-std=c++14"],
+                extra_compile_args=["-std=c++14"] if os.name != 'nt' else ['/std:c++14', '/MP'],
+                extra_link_args=["-std=c++14" if os.name != 'nt' else '/std:c++14'],
```
Comment (peterjc123) [inline setup.py:359]: The flag is useless at linking stage for MSVC.

**Review by peterjc123 — COMMENTED** [review]
Generally LGTM, just a few inline comments

File: packaging/pkg_helpers.bash (hunk 2 above)
Comment (peterjc123) [reply packaging/pkg_helpers.bash:232]: Okay, I think I know the reason. We are using `ffmpeg` from the pytorch channel.

**Comment by andfoy** [comment]
It seems that disabling MMX also fixes the segfaults on Windows, but some test errors remain

File: packaging/pkg_helpers.bash (hunk 1 above)
Comment (andfoy) [reply packaging/pkg_helpers.bash:312]: I think we can remove the version pinning safely

**Comment by andfoy** [comment]
Update: MMX disables also SSE, so basically we're losing all hardware-specific optimizations. It seems that the error is related to [YUV to RGB](https://github.com/FFmpeg/FFmpeg/blob/161495a3d026a73bc802c66a94e867ef457320d4/libswscale/x86/yuv_2_rgb.asm#L376) conversion, invoked underneath https://github.com/pytorch/vision/blob/c36dc43b115b495943813a26004c740d8bffd5b6/torchvision/csrc/cpu/decoder/video_sampler.cpp#L45

**Comment by andfoy** [comment]
After speaking with @bjuncek, we found out that the segfaults are related to this issue: https://github.com/pytorch/vision/issues/2650

**Comment by andfoy** [comment]
FFmpeg 4.2 works with this PR at least on Linux (and possibly on Mac). After discussing with @fmassa, we decided to disable FFmpeg on Windows whilist https://github.com/pytorch/vision/issues/2688 is fixed

**Comment by fmassa** [comment]
@andfoy tests are segfaulting in video-reading functions

**Comment by andfoy** [comment]
I think 4.2 packages are not ready in builder

**Comment by fmassa** [comment]
Interesting, test failures could seem related, although most of them are on `test_io.py` which uses pyav, so I'm not 100% sure what's happening here. Maybe pyav is picking a different ffmpeg version than the one it was compiled to use?

For the error in `test_read_video_from_memory_scripted`, there is an easy fix: replace the `%` for string formatting in
https://github.com/pytorch/vision/blob/8dfcff745a5bd2d4886716bf0deff8dcc8e75fed/torchvision/io/_video_opt.py#L91
with the `.format` equivalent. There might have been a change in torchscript that made the `%` stop working.

**Comment by andfoy** [comment]
The error here is that some tests are trying to use `x264` and `x264rgb`, which are not available on our LGPL FFmpeg distribution. I tried to switch some of them to `h264`, but it fails when `x264rgb` is required

**Comment by andfoy** [comment]
For some reason, tests pass when `test_io.py` are executed with `pytest -v test/test_io.py`

imagen

However, it fails when is ran alongside all the testsuite (`pytest -v test`)

imagen

**Comment by fmassa** [comment]
@andfoy if tests pass when only running on `test_io`, but not the full test suite, this might indicate that another test (`test_video_reader`?) is also loading ffmpeg from our binaries and affects `test_io`. To make sure that this is the case, can you run `test_io` and `test_video_reader` in the same run to double-check (although the order on which the tests are run might be important)

**Comment by andfoy** [comment]
Found more information related to the test failure cases:

https://github.com/pytorch/vision/pull/2745#issuecomment-703937446

**Comment by fmassa** [comment]
Great, this is awesome, we will soon be unblocking this and @bjuncek PR I believe

File: packaging/pkg_helpers.bash (hunk 2 above)
Comment (fmassa) [reply packaging/pkg_helpers.bash:243]: Just to be sure, this means that if there is a package under `pytorch-nightly` and `pytorch`, the one from `pytorch-nightly` is picked as it comes first in the priority, right?

**Review by fmassa — APPROVED** [review]
Let's merge this PR and see how it goes.

We can clean a few things up (following @peterjc123 comment) on a follow-up PR.

Also, ccing @malfet and @seemethere for visibility -- we are adding ffmpeg as a dependency for torchvision.

**Pull request merged**